### PR TITLE
[UCX] Add UCX support

### PR DESCRIPTION
**Pull request by mengdiz97**
**Description of changes:**

- [x] Add UCX support for the build of FlexFlow
- [x] Add corresponding tests
- [x] Update documentation

**Related Issues:**

Linked Issues:
- Issue #

Issues closed by this PR:
- Closes #

**Before merging:**

- [ ] Did you update the [flexflow-third-party](https://github.com/flexflow/flexflow-third-party) repo, if modifying any of the Cmake files, the build configs, or the submodules?

### CODE REVIEW

**Comment by jiazhihao** [comment]
@gabrieleoliaro Do you think we can merge this PR now? Or should we wait for the Python inference fix first?

File: scripts/multinode_run.sh [hunk 1]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+eval "$(conda shell.bash hook)"
+conda activate flexflow
+~/FlexFlow/python/flexflow_python ~/FlexFlow/examples/python/native/mnist_mlp.py -ll:py 1 -ll:gpu 1 -ll:fsize 8000 -ll:zsize 8000
```
Comment (gabrieleoliaro) [inline scripts/multinode_run.sh:4]: newline here

File: CMakeLists.txt [hunk 2]
```diff
@@ -46,9 +46,93 @@ option(FF_USE_ALL_PREBUILT_LIBRARIES "Enable use of all pre-compiled libraries,
 # option for using Python
 option(FF_USE_GASNET "Run FlexFlow with GASNet" OFF)
 set(FF_GASNET_CONDUITS aries udp mpi ibv ucx)
-set(FF_GASNET_CONDUIT "mpi" CACHE STRING "Select GASNet conduit ${FF_GASNET_CONDUITS}")
+set(FF_GASNET_CONDUIT ${FF_GASNET_CONDUIT} CACHE STRING "Select GASNet conduit ${FF_GASNET_CONDUITS}")
```
Comment (gabrieleoliaro) [inline CMakeLists.txt:49]: why do we need this change?

File: scripts/mnist_mlp_run.sh [hunk 3]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+eval "$(conda shell.bash hook)"
+conda activate flexflow
+~/FlexFlow/python/flexflow_python ~/FlexFlow/examples/python/native/mnist_mlp.py -ll:py 1 -ll:gpu 1 -ll:fsize 8000 -ll:zsize 8000
```
Comment (gabrieleoliaro) [inline scripts/mnist_mlp_run.sh:4]: looks like we need a newline here too

**Review by gabrieleoliaro — CHANGES_REQUESTED** [review]
left two more comments :) also, could you resolve the merge conflict, so we can merge the PR as soon as the CI tests pass?

File: CMakeLists.txt (hunk 2 above)
Comment (mengdiz97) [reply CMakeLists.txt]: I supposed we need to pass the variable to here. It is not necessary and I have reverted it.

**Review by gabrieleoliaro — APPROVED** [review]

**Comment by eddy16112** [comment]
Have you tried to run flexflow with the legion UCX backend, not the gasnet backend with UCX? Since legion now supports native UCX , I do not think it is necessary to use the gasnet with UCX.

**Comment by eddy16112** [comment]
BTW, to enable the legion UCX backend, you can set the Legion_NETWORKS to ucx. https://github.com/flexflow/FlexFlow/blob/master/cmake/legion.cmake#L132

**Pull request merged**